portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java

```diff
@@ -0,0 +1,91 @@
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+
+import java.lang.System;
+
+@WebServlet("/user/*")
+public final class AuthenticationServlet extends HttpServlet {
+    private UserService userService = UserServiceFactory.getUserService();
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String uri = request.getRequestURI(); // see below - we'll be removed after the review
+
+        switch(this.getActionPath(request.getRequestURI())) { // REVIEW: is this acceptable? or shall I use uri above?
+            case "login":
+                this.doLogin(request, response);
+                break;
+
+            case "logout":
+                this.doLogout(request, response);
+                break;
+
+            case "data":
+                this.doGetData(request, response);
+                break;
+
+            default:
+                this.doHandleDefault(request, response);
+        }
+    }
+
+    private void doLogin(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        final String redirectUrl = request.getContextPath() + "/";
+
+        if(this.userService.isUserLoggedIn()) {
+            response.sendRedirect(redirectUrl);
+        } else {
+            response.sendRedirect(this.userService.createLoginURL(redirectUrl));
+        }
+    }
+
+    private void doLogout(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        final String redirectUrl = request.getContextPath() + "/";
+
+        if(this.userService.isUserLoggedIn()) {
+            response.sendRedirect(this.userService.createLogoutURL(redirectUrl));
+        } else {
+            response.sendRedirect(request.getContextPath() + "/");
+        }
+    }
+
+    private void doGetData(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("text/plain;");
+
+        if(this.userService.isUserLoggedIn()) {
+            response.getWriter().write(this.userService.getCurrentUser().getEmail());
+        } else {
+            response.getWriter().write("false");
+        }
+    }
+
+    private void doHandleDefault(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.sendRedirect(request.getContextPath() + "/");
+    }
+
+    /**
+     * Extracts the type of action needed to be taken based on the URI.
+     * @param uri the page's uri in the form /user/[action]
+     * @return [action] if it does not contain any slashes, otherwise an empty string
+     */
+    private String getActionPath(String uri) {
+        uri = uri.substring(1); // discard the leftmost slash
+
+        String []splittedBySlashes = uri.split("/");
+
+        if(splittedBySlashes.length >= 3 || splittedBySlashes.length <= 1) { // not acceptable number of overall slashes
```

Comment:
s/not acceptable/unacceptable/

But why is this the case?

Reply:
Because we need action not to have any slashes (and also not to be empty). In such a case, because we split it by a '/', if it's empty or has at least one such slash, then the total length of the splitBySlashes will be grater than two. The uri will be:
"user/[action]", where [action] is our action.

Reply:
This  should be explained in comments as it seems mysterious